src/tools/aml-language-server/handlers.ts

```diff
@@ -0,0 +1,141 @@
+/**
+ * @license
+ * Copyright (c) 2019 Google Inc. All rights reserved.
+ * This code may only be used under the BSD style license found at
+ * http://polymer.github.io/LICENSE.txt
+ * Code distributed by Google as part of this project is also
+ * subject to an additional IP rights grant found at
+ * http://polymer.github.io/PATENTS.txt
+ */
+
+import {Message} from 'vscode-jsonrpc';
+import {NotificationMessage} from 'vscode-jsonrpc/lib/messages';
+import {CodeActionParams, Command, CompletionItemKind, CompletionList, Diagnostic, DiagnosticSeverity, DidChangeConfigurationParams, DidChangeTextDocumentParams, DidCloseTextDocumentParams, DidOpenTextDocumentParams, DidSaveTextDocumentParams, DocumentSymbolParams, ExecuteCommandParams, Hover, InsertTextFormat, Location, MarkedString, ParameterInformation, PublishDiagnosticsParams, Range, ReferenceParams, RenameParams, SignatureHelp, SignatureInformation, SymbolInformation, TextDocumentPositionParams, TextDocumentSyncKind, TextEdit, WorkspaceEdit} from 'vscode-languageserver';
+
+import {Dictionary} from '../../runtime/hot.js';
+import {Manifest, ManifestError} from '../../runtime/manifest.js';
+
+import {LspLoader} from './lspLoader.js';
+import * as MessageTypes from './messageTypes.js';
+import {alphaNumerics, AmlServiceContext, jsonrpc, normalizeUri, uri2path} from './util.js';
+
+// tslint:disable-next-line: no-any
+export type Handler = ((params: object, context: AmlServiceContext) => any);
+
+export const handlers: Dictionary<Handler> = {
+  initialize: (params: object, context: AmlServiceContext):
+                  MessageTypes.InitializeResult => {
+    context.rootPath =
+    // tslint:disable-next-line: no-any
+        (params as any).rootPath || uri2path((params as any).rootUri);
+    const result: MessageTypes.InitializeResult = {
+      capabilities: {
+        // Tell the client that the server works in FULL text document sync mode
+        textDocumentSync: TextDocumentSyncKind.Full,
+        hoverProvider: true,
+        signatureHelpProvider: {
+          triggerCharacters: ['(', ','],
+        },
+        definitionProvider: true,
+        typeDefinitionProvider: true,
+        referencesProvider: true,
+        documentSymbolProvider: true,
+        workspaceSymbolProvider: true,
+        xworkspaceReferencesProvider: true,
+        xdefinitionProvider: true,
+        xdependenciesProvider: true,
+        completionProvider: {
+          resolveProvider: true,
+          triggerCharacters: ['.'],
+        },
+        codeActionProvider: true,
+        renameProvider: true,
+        executeCommandProvider: {
+          commands: [],
+        },
+        xpackagesProvider: true,
+      }
+    };
+    return result;
+  },
+
+  textDocumentCompletion: async (
+      params, context: AmlServiceContext):
+      Promise<MessageTypes.CompletionItem> => {
+        const uri = (params as {textDocument: {uri: string}}).textDocument.uri;
```

Comment:
Assuming `params` is set up as `any` to begin with (as commented above), this could just be `const uri = params.textDocument.uri as string;` - same type & object structure info, more concise.

Reply:
Done